paddle/fluid/operators/spectral_norm_op.cc

```diff
@@ -0,0 +1,180 @@
+   limitations under the License. */
+
+#include "paddle/fluid/operators/spectral_norm_op.h"
+#include "paddle/fluid/framework/op_registry.h"
+
+namespace paddle {
+namespace operators {
+
+using framework::Tensor;
+
+class SpectralNormOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+ protected:
+  void InferShape(framework::InferShapeContext* ctx) const override {
+    PADDLE_ENFORCE(ctx->HasInput("Weight"),
+                   "Input(Weight) of SpectralNormOp should not be null.");
+    PADDLE_ENFORCE(ctx->HasInput("U"),
+                   "Input(U) of SpectralNormOp should not be null.");
+    PADDLE_ENFORCE(ctx->HasInput("V"),
+                   "Input(V) of SpectralNormOp should not be null.");
+    PADDLE_ENFORCE(ctx->HasOutput("Out"),
+                   "Output(Out) of SpectralNormOp should not be null.");
+
+    auto dim_weight = ctx->GetInputDim("Weight");
+    auto rank_weight = dim_weight.size();
+    PADDLE_ENFORCE(rank_weight >= 2 && rank_weight <= 5,
+                   "The rank of Input(Weights) can only be 2, 3,"
+                   "4, 5 for fc, conv1d, conv2d, conv3d layers.");
+
+    int dim = ctx->Attrs().Get<int>("dim");
+    int power_iters = ctx->Attrs().Get<int>("power_iters");
+    PADDLE_ENFORCE(dim == 0 || dim == 1, "Attr(dim) can only be 0 or 1");
+    PADDLE_ENFORCE(power_iters >= 0,
+                   "Attr(power_iters) should be larger equal then 0");
+
+    int h = dim_weight[dim];
+    int w = 1;
+    for (int i = 0; i < rank_weight; i++) {
+      if (i != dim) {
+        w *= dim_weight[i];
+      }
+    }
+    auto dim_u = ctx->GetInputDim("U");
+    auto dim_v = ctx->GetInputDim("V");
+    PADDLE_ENFORCE_EQ(dim_u[0], h,
+                      "Input(U) dims[0] should be equal to "
+                      "Input(Weight) dims[Attr(dim)]");
+    PADDLE_ENFORCE_EQ(
+        dim_v[0], w,
+        "Input(V) dims[0] should be equal to "
+        "the product of Input(Weight) dims except dims[Attr(dim)]");
+
+    ctx->SetOutputDim("Out", dim_weight);
+    ctx->ShareLoD("Weight", /*->*/ "Out");
+  }
+
+ protected:
+  framework::OpKernelType GetExpectedKernelType(
+      const framework::ExecutionContext& ctx) const override {
+    return framework::OpKernelType(ctx.Input<Tensor>("Weight")->type(),
+                                   ctx.GetPlace());
+  }
+};
+
+class SpectralNormOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() override {
+    AddInput("Weight",
+             "The input weight tensor of spectral_norm operator, "
+             "This can be a 2-D, 3-D, 4-D, 5-D tensor which is the"
+             "weights of fc, conv1d, conv2d, conv3d layer.");
+    AddInput("U",
+             "The weight_u tensor of spectral_norm operator, "
+             "This can be a 1-D tensor in shape [H, 1],"
+             "H is the 1st dimentions of Weight after reshape"
```

Comment:
mention the shape after reshape is better

Reply:
Done